services/src/main/java/org/exoplatform/officeonline/WOPILockManagerPlugin.java

```diff
@@ -0,0 +1,236 @@
+  public void unlock(Node node, String lockId, String workspace) throws RepositoryException, LockMismatchException {
+    FileLock fileLock = locks.get(node.getUUID());
+    if (fileLock != null) {
+      if (lockId.equals(fileLock.getLockId())) {
+        getUserSession(workspace).addLockToken(fileLock.getLockToken());
+        node.unlock();
+        locks.remove(node.getUUID());
+        LOG.debug("Lock removed from node UUID: {}, lockId: {}", node.getUUID(), lockId);
+      } else {
+        throw new LockMismatchException("Provided lockId doesn't match lock on the file", fileLock.getLockToken());
+      }
+    } else {
+      throw new LockMismatchException("File isn't locked", "");
+    }
+  }
+
+  /**
+   * Refresh lock.
+   *
+   * @param node the node
+   * @param lockId the lock id
+   * @throws RepositoryException the repository exception
+   * @throws LockMismatchException the lock mismatch exception
+   */
+  public void refreshLock(Node node, String lockId) throws RepositoryException, LockMismatchException {
+    FileLock fileLock = locks.get(node.getUUID());
+    if (fileLock != null) {
+      if (lockId.equals(fileLock.getLockId())) {
+        fileLock.setExpires(System.currentTimeMillis() + LOCK_EXPIRES);
+        LOG.debug("Lock refreshed. UUID: {}, lockId: {}", node.getUUID(), lockId);
+      } else {
+        throw new LockMismatchException("Provided lockId doesn't match lock on the file", fileLock.getLockToken());
+      }
+    } else {
+      throw new LockMismatchException("File isn't locked", "");
+    }
+  }
+
+  /**
+   * Gets the user session.
+   *
+   * @param workspace the workspace
+   * @return the user session
+   * @throws RepositoryException the repository exception
+   */
+  protected Session getUserSession(String workspace) throws RepositoryException {
+    if (workspace == null) {
+      workspace = jcrService.getCurrentRepository().getConfiguration().getDefaultWorkspaceName();
+    }
+    SessionProvider sp = sessionProviders.getSessionProvider(null);
+    return sp.getSession(workspace, jcrService.getCurrentRepository());
+  }
+
+  /**
+   * Gets the system session.
+   *
+   * @return the system session
+   * @throws RepositoryException the repository exception
+   */
+  protected Session getSystemSession() throws RepositoryException {
+    String workspace = jcrService.getCurrentRepository().getConfiguration().getDefaultWorkspaceName();
+    SessionProvider sp = SessionProvider.createSystemProvider();
+    return sp.getSession(workspace, jcrService.getCurrentRepository());
+  }
+
+  /**
+   * Removes the expired.
+   */
+  protected void removeExpired() {
+    List<? extends FileLock> fileLocks = new ArrayList<FileLock>();
+    try {
+      fileLocks = locks.getCachedObjects();
+    } catch (Exception e) {
+      LOG.error("Cannot get cached locks from ExoCache", e);
+      return;
+    }
+   
+    fileLocks.forEach(lock -> {
+      if ((lock.getExpires() - System.currentTimeMillis()) < EXPIRES_DELAY) {
```

Comment:
We could use the cache expiration and then use CacheListener.onExpire(), a key (fileId) will be given in parameters.

Reply:
Used selector for removing expired locks.

Reply:
But why cache expiration doesn't fit for us? We should lock on exact lock time, can it be possible the expiration time will be configurable by a customer?

Reply:
1) We have logic with a refreshing expiration time (REFRESH_LOCK), so the current approach is more convinient. 
2) The expiration time must be 30 minutes according to WOPI docs: https://wopi.readthedocs.io/projects/wopirest/en/latest/concepts.html#term-lock

Reply:
OK, we should be locked on exact time value for WOPI.